tests/transform_dialect/cuda/mma.mlir

```diff
@@ -0,0 +1,69 @@
+// RUN: iree-opt %s --iree-transform-dialect-interpreter --transform-dialect-drop-schedule --split-input-file | FileCheck %s
+
+#matmat_accesses = [
+  affine_map<(m, n, k) -> (m, k)>,
+  affine_map<(m, n, k) -> (k, n)>,
+  affine_map<(m, n, k) -> (m, n)>
+]
+#matmat_trait = {
+  indexing_maps = #matmat_accesses,
+  iterator_types = ["parallel", "parallel", "reduction"]
+}
+
+func.func @wmma(%a: memref<16x16xf32>, %b: memref<16x16xf32>, %c: memref<16x16xf32>) {
+  %c0 = arith.constant 0: index
+  %cst = arith.constant 0.0: f32
+  %va = vector.transfer_read %a[%c0, %c0], %cst: memref<16x16xf32>, vector<16x16xf32>
+  %vb = vector.transfer_read %b[%c0, %c0], %cst: memref<16x16xf32>, vector<16x16xf32>
+  %vc = vector.transfer_read %c[%c0, %c0], %cst: memref<16x16xf32>, vector<16x16xf32>
+
+  // CHECK-NOT: vector.contract
+  //     CHECK:  gpu.subgroup_mma_compute
+  %vres = vector.contract #matmat_trait %va, %vb, %vc
+    : vector<16x16xf32>, vector<16x16xf32> into vector<16x16xf32>
+  vector.transfer_write %vres, %c[%c0, %c0]: vector<16x16xf32>, memref<16x16xf32>
+  return
+}
+
+transform.sequence failures(propagate) {
+^bb1(%module: !pdl.operation):
+  %func = transform.structured.match ops{["func.func"]} in %module
+    : (!pdl.operation) -> !pdl.operation
+  %func_2 = transform.iree.apply_patterns %func { unroll_vectors_gpu_wmma }
+  transform.iree.vector.vector_to_mma_conversion %func_2 { use_wmma }
+}
+
+// -----
+
+#matmat_accesses = [
+  affine_map<(m, n, k) -> (m, k)>,
+  affine_map<(m, n, k) -> (k, n)>,
+  affine_map<(m, n, k) -> (m, n)>
+]
+#matmat_trait = {
+  indexing_maps = #matmat_accesses,
+  iterator_types = ["parallel", "parallel", "reduction"]
+}
+
+func.func @mma_sync(%a: memref<16x16xf32>, %b: memref<16x16xf32>, %c: memref<16x16xf32>) {
+  %c0 = arith.constant 0: index
+  %cst = arith.constant 0.0: f32
+  %va = vector.transfer_read %a[%c0, %c0], %cst: memref<16x16xf32>, vector<16x16xf32>
+  %vb = vector.transfer_read %b[%c0, %c0], %cst: memref<16x16xf32>, vector<16x16xf32>
+  %vc = vector.transfer_read %c[%c0, %c0], %cst: memref<16x16xf32>, vector<16x16xf32>
+
+  // CHECK-NOT: vector.contract
+  //     CHECK: nvgpu.mma.sync
+  %vres = vector.contract #matmat_trait %va, %vb, %vc
+    : vector<16x16xf32>, vector<16x16xf32> into vector<16x16xf32>
+  vector.transfer_write %vres, %c[%c0, %c0]: vector<16x16xf32>, memref<16x16xf32>
+  return
+}
+
+transform.sequence failures(propagate) {
+^bb1(%module: !pdl.operation):
+  %func = transform.structured.match ops{["func.func"]} in %module
+    : (!pdl.operation) -> !pdl.operation
+  %func_2 = transform.iree.apply_patterns %func { unroll_vectors_gpu_mma_sync }
+  transform.iree.vector.vector_to_mma_conversion %func_2 { use_mma_sync }
```

Comment:
In the current implementation, I went for a separation of the (step 1) `unrolling` which takes vector-dialect IR and produces vector-dialect IR, from the (step 2) vector_to_mma rewrite which takes vector dialect and produces gpu-specific dialect.

This is partly biased by the fact that I understand (step 1) well and did want to be able to look at the IR (i.e. by commenting out l68-end and running the script) without jumping an abstraction gap that I did not yet fully understand (step 2).

I do generally find the ability to just comment script lines out and immediately see the produced IR I want (without recompiling the compiler and/or sifting through many lines of dump), very valuable for velocity and building understanding.

This is all subject to preference and customization and different people's taste will vary, so I am very much open to making simplifying changes.

Reply:
Ah yes, there is another aspect: I am also separating the pure vector-dialect transforms from the gpu rewrite because as I comment [here](https://github.com/iree-org/iree-samples/blob/main/transform_dialect/examples/cuda/matmul_codegen_spec_step_02_pad_shared_wmma.mlir#L122) step 1 can happen on tensors while step 2 must happen on buffers to benefit from the special layout hiding memory types for WMMA ops.

Therefore, the conversion to GPU introduces phase-ordering constraints that I want to be able to illustrate and reason about.